land-of-wizards/core.clj

```diff
@@ -0,0 +1,57 @@
+(ns land-of-wizards.core
+  (:require [clojure.string :as str]
+            [land-of-wizards.data :refer :all]))
+
+(comment "The first command to have is a command that tells about the location the wizard is standing in. So what would a function need to describe a location in a world? Well, it would need to know the location to describe and would need to be able to look at a map and find that location on the map. Here's a function, that does exactly that")
+
+(defn describe-location [location nodes]
+  (get nodes location))
+
+#_ (describe-location :living-room nodes)
+
+(comment "The `describe-location` seems pretty awkward in several different ways. First of all, why are the variables for location and map being passed as parameters, instead of just reading the global variables directly? The reason is that Lispers often like to write code in the Functional Programming Style (To be clear, this is completely unrelated in any way to the concept called \"procedural programming\" or \"structural programming\"). In this style, the goal is to write functions that are referentially transaparent. An expression is called referentially transparent if it can be replaced with its corresponding value without changing the program's behavior, and to achive that:
+
+1.  You only read variables that are passed into the function or are created by the function (So you don't read any global variables)
+2.  You never change the value of a variable that has already been set (So no incrementing variables or other such foolishness)
+3.  You never interact with the outside world, besides returning a result value. (So no writing to files, no writing messages for the user)")
+
+(comment "Here's a function to describe the path from the wizard's current location to any given location")
+
+(defn describe-path [edges from to]
+  (let [[path obstacle] (to (from edges))]
+    (str "There is a " (name obstacle) " going "
+         (name path) " from here.")))
+
+#_ (describe-path edges :living-room :attic)
+
+(comment "Now there's a function to describe all paths in and out of wizard's current location. This function uses another common functional programming technique: The use of Higher Order Functions - This means that the map function is taking other functions as parameters so that they can call them themselves - map simply applies another function to every object in the collection (in this case a list) returned by `(keys (location edges))`, basically causing all paths to be changed into pretty descriptions procided by the describe-path function.")
+
+(defn describe-all-paths [edges location]
+  (str/join " " (map #(describe-path edges location %) (keys (location edges)))))
```

Comment:
- `(get edges location)` might be more suitable here(?)

- Would we want to introduce apply here?

```clj
(defn describe-all-paths [edges location]
  (apply str (map #(describe-path edges location %) (keys (get edges location)))))
```
- Also I couldn't help but write this (pl ignore if threading feels like too much):
```clj
(defn describe-all-paths [edges location]
  (->> location
       (get edges)
       keys
       (map #(describe-path edges location %))
       (apply str)))
```

Reply:
> (keys (get edges location)) might be more suitable here(?) 

It's worth introducing lookup by key and the concept of invoking keywords as functions since it's intrinsic to Clojure's data structures. (The Extraction section under Maps in content.org introduces `get` along with other ways to do a lookup).

> Would we want to introduce apply here?

I've consciously tried to not use apply (and several other functions). After looking at the existing content on ClojureBridge-blr it occurred that a lot of functions/concepts (including `apply`) might have been out of the scope of this workshop the last time it was held.

> Also I couldn't help but write this (pl ignore if threading feels like too much 😛):

Almost the same rationale as with not using `apply`. I don't think we should be introducing macros at all unless we also want to delve into what macros are.